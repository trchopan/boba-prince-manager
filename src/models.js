export const profileParser = data => ({
  fullname: data.fullname || "",
  position: data.position || "user",
  avatar: data.avatar || "/images/bloody-smile.jpg"
});
