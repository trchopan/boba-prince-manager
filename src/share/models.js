export const themes = {
  light: "Light",
  dark: "Dark"
};

export const languages = {
  vi: { code: "vi", name: "Tiáº¿ng Viá»‡t", char: "ðŸ‡»ðŸ‡³" },
  en: { code: "en", name: "English", char: "ðŸ‡¬ðŸ‡§" }
};

export const profileParser = data => ({
  fullname: data.fullname || "",
  position: data.position || "user",
  avatar: data.avatar || "/images/bloody-smile.jpg"
});
