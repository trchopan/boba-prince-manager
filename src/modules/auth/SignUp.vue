<template>
  <v-layout column>
    <v-layout
      row
      :wrap="true"
      align-center
      justify-center
    >
      <v-flex
        xs12
        sm6
        md4
      >
        <v-card
          key="profile-edit"
          class="elevation-6 pt-3 px-3 text-xs-center"
        >
          <v-form
            ref="signUpForm"
            v-model="signUpFormValid"
            @submit.prevent="signUp()"
          >
            <v-text-field
              autofocus
              prepend-icon="person"
              label="Email"
              type="text"
              v-model="email"
              :rules="emailRules"
              validate-on-blur
              required
            />
            <v-text-field
              prepend-icon="lock"
              :label="$t.password"
              type="password"
              v-model="password"
              :rules="passwordRules"
              validate-on-blur
              required
            />
            <v-text-field
              prepend-icon="lock"
              :label="$t.passwordConfirm"
              type="password"
              v-model="passwordConfirm"
              :rules="passwordConfirmRules"
              validate-on-blur
              required
            />
            <Recaptcha @response="recaptchaResponse = $event"/>
            <span
              class="warn--text"
              v-if="authError"
            >{{ authError }}</span>
            <v-card-actions>
              <v-spacer/>
              <v-btn
                outline
                color="secondary"
                type="submit"
                :disabled="loading.create"
              >{{ $t.signUp }}</v-btn>
            </v-card-actions>
          </v-form>
        </v-card>
      </v-flex>
    </v-layout>
  </v-layout>
</template>

<script>
import { mapGetters } from "vuex";
import { validateEmail } from "@/helpers.js";
import Recaptcha from "./components/Recaptcha.vue";

const languagesMap = {
  password: { vi: "Mật khẩu", en: "Password" },
  passwordConfirm: { vi: "Xác nhận mật khẩu", en: "Confirm Password" },
  signUp: { vi: "Đăng ký", en: "Sign Up" },
  invalidEmail: {
    vi: "Email không hợp lệ",
    en: "Invalid email"
  },
  passwordGreaterThan: {
    vi: "Mật khẩu phải nhiều hơn {0} ký tự",
    en: "Password must have more than {0} characters"
  },
  passwordNotMatch: {
    vi: "Mật khẩu xác nhận không trùng nhau",
    en: "Password confirmation does not match"
  },
  requireCaptcha: {
    vi: "Cần xác thực captcha",
    en: "Require captcha verification"
  },
  emailAlreadyExist: {
    vi: "Đã có tài khoản sử dụng email này",
    en: "Email is already used by another account"
  }
};

export default {
  name: "Login",
  components: {
    Recaptcha
  },
  data() {
    return {
      signUpFormValid: false,
      email: "",
      password: "",
      passwordConfirm: "",
      emailRules: [v => validateEmail(v) || this.$t.invalidEmail],
      passwordRequiredLength: 6,
      passwordRules: [
        v =>
          v.length >= this.passwordRequiredLength ||
          this.$format(this.$t.passwordGreaterThan, [
            this.passwordRequiredLength
          ])
      ],
      passwordConfirmRules: [
        v => v === this.password || this.$t.passwordNotMatch
      ]
    };
  },
  computed: {
    ...mapGetters({
      error: "error",
      authUser: "auth/authUser",
      loading: "auth/loading",
      language: "layout/language"
    }),
    $t() {
      return this.$translate(languagesMap, this.language.value);
    },
    authError() {
      const lastError = this.error[this.error.length - 1] || null;
      if (!lastError) {
        return "";
      }
      switch (lastError.code) {
        case "auth/require-captch":
          return this.$t.requireCaptcha;
        case "auth/email-already-exists":
          return this.$t.emailAlreadyExist;
        default:
          return this.$t.unknownError;
      }
    }
  },
  methods: {
    async signUp() {
      this.$refs.signUpForm.validate();
      if (this.signUpFormValid) {
        const success = await this.$store.dispatch("auth/createUser", {
          email: this.email,
          password: this.password,
          response: this.recaptchaResponse
        });

        if (success) {
          this.$router.replace("/login");
        }
      }
    }
  },
  mounted() {
    if (this.authUser) {
      this.$router.replace("/");
    }
  },
  beforeRouteLeave(to, from, next) {
    this.$store.dispatch("clearError");
    next();
  }
};
</script>

<style lang="scss" scoped>
.v-progress-linear {
  margin: 0;
}
</style>
