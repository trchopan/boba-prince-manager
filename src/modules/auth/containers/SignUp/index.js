import SignUp from "./SignUp.vue";

export { SignUp };
