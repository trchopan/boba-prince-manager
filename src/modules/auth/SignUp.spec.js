import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import Translation from "@/plugins/translation.js";
import SignUp from "./SignUp.vue";
import { mockCustomElements } from "@/__mocks__/custom-elements.js";
import { languages } from "@/modules/core/layout.models.js";

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(Translation);

describe("SignUp.vue", () => {
  let store;
  let getters;
  let actions;
  let wrapper;

  beforeEach(() => {
    getters = {
      error: () => [],
      "layout/language": () => languages.en,
      "auth/authUser": () => null,
      "auth/loading": () => ({ create: false })
    };
    actions = {
      "auth/createUser": jest
        .fn()
        .mockImplementationOnce(() => Promise.resolve(false))
        .mockImplementationOnce(() => Promise.resolve(true))
    };
    store = new Vuex.Store({
      getters,
      actions
    });
    let stubs = {};
    mockCustomElements.forEach(x => {
      stubs[x] = { render: () => x };
    });
    wrapper = shallowMount(SignUp, {
      stubs: mockCustomElements,
      mocks: { $router: { replace: jest.fn() } },
      store,
      localVue
    });
  });

  it("signs up", async () => {
    wrapper.vm.$refs.signUpForm = { validate: jest.fn() };
    wrapper.vm.signUpFormValid = false;

    await wrapper.vm.signUp();
    expect(wrapper.vm.$refs.signUpForm.validate).toBeCalled();
    expect(wrapper.vm.$router.replace).not.toBeCalled();

    wrapper.vm.email = "test@test.com";
    wrapper.vm.password = "123456";
    wrapper.vm.recaptchaResponse = "abcxyz123";
    wrapper.vm.signUpFormValid = true;

    await wrapper.vm.signUp(); // auth/createUser resolve false
    expect(actions["auth/createUser"].mock.calls[0][1]).toEqual({
      email: wrapper.vm.email,
      password: wrapper.vm.password,
      response: wrapper.vm.recaptchaResponse
    });
    expect(wrapper.vm.$router.replace).not.toBeCalled();

    await wrapper.vm.signUp(); // auth/createUser resolve true
    expect(wrapper.vm.$router.replace.mock.calls).toEqual([["/login"]]);
  });

  it("matchs snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
