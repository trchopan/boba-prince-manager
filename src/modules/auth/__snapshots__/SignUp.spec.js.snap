// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp.vue matchs snapshot 1`] = `
<v-layout-stub column="">
  <v-layout-stub row="" wrap="true" align-center="" justify-center="">
    <v-flex-stub xs12="" sm6="" md4="">
      <v-card-stub class="elevation-6 pt-3 px-3 text-xs-center">
        <v-form-stub>
          <v-text-field-stub autofocus="" prepend-icon="person" label="Email" type="text" rules="v => (0, _helpers.validateEmail)(v) || this.$t.invalidEmail" validate-on-blur="" required=""></v-text-field-stub>
          <v-text-field-stub prepend-icon="lock" label="Password" type="password" rules="v => v.length >= this.passwordRequiredLength || (0, _helpers.format)(this.$t.passwordGreaterThan, [this.passwordRequiredLength])" validate-on-blur="" required=""></v-text-field-stub>
          <v-text-field-stub prepend-icon="lock" label="Confirm Password" type="password" rules="v => v === this.password || this.$t.passwordNotMatch" validate-on-blur="" required=""></v-text-field-stub>
          <recaptcha-stub></recaptcha-stub>
          <!---->
          <v-card-actions-stub>
            <v-spacer-stub></v-spacer-stub>
            <v-btn-stub outline="" color="secondary" type="submit">Sign Up</v-btn-stub>
          </v-card-actions-stub>
        </v-form-stub>
      </v-card-stub>
    </v-flex-stub>
  </v-layout-stub>
</v-layout-stub>
`;
